--- a/config.def.h
+++ b/config.def.h
@@ -11,9 +11,9 @@
 
 static const char *colorname[NUMCOLS] = {
 	[BACKGROUND] =   "black",     /* after initialization */
-	[INIT] =   "#2d2d2d",     /* after initialization */
-	[INPUT] =  "#005577",   /* during input */
-	[FAILED] = "#CC3333",   /* wrong password */
+	[INIT] =   "#141a1b",     /* after initialization */
+	[INPUT] =  "#2eb398",   /* during input */
+	[FAILED] = "#cc0000",   /* wrong password */
 };
 
 /* treat a cleared input like a wrong password (color) */
