--- a/config.def.h	2019-06-26 15:12:55.579092634 -0700
+++ b/config.def.h      2019-06-26 15:16:01.603309778 -0700
@@ -5,13 +5,13 @@
 static const unsigned int snap      = 32;       /* snap pixel */
 static const int showbar            = 1;        /* 0 means no bar */
 static const int topbar             = 1;        /* 0 means bottom bar */
-static const char *fonts[]          = { "monospace:size=10" };
-static const char dmenufont[]       = "monospace:size=10";
-static const char col_gray1[]       = "#222222";
+static const char *fonts[]          = { "noto sans:size=10" };
+static const char dmenufont[]       = "noto sans:size=10";
+static const char col_gray1[]       = "#2f343f";
 static const char col_gray2[]       = "#444444";
-static const char col_gray3[]       = "#bbbbbb";
-static const char col_gray4[]       = "#eeeeee";
-static const char col_cyan[]        = "#005577";
+static const char col_gray3[]       = "#dddddd";
+static const char col_gray4[]       = "#ffffff";
+static const char col_cyan[]        = "#5294e2";
 static const char *colors[][3]      = {
 	/*               fg         bg         border   */
 	[SchemeNorm] = { col_gray3, col_gray1, col_gray2 },
